icu4c/source/common/rbbirb.cpp

```diff
@@ -302,13 +313,15 @@ RBBIDataHeader *RBBIRuleBuilder::build(UErrorCode &status) {
     }
 #endif
 
-    fSetBuilder->buildTrie();
+    UBool use8Bits = fSetBuilder->getNumCharCategories() < 127;
```

Comment:
Also need to check that the number of states (table rows) fits in a byte for 8 bit tables.